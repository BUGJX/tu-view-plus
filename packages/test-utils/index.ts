export { default as defineGetter } from './define-getter';
