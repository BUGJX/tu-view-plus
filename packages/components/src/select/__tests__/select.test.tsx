import { mount } from '@vue/test-utils'
import { describe, expect, test } from 'vitest'
import Select from '../src/select.vue'

describe('Select.vue', () => {
  test('show dropdown correctly', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Option1', 'Option2', 'Option3', 'Option4'],
      },
    });

    await wrapper.find('.tu-select-view').trigger('click');

    expect(document.body.outerHTML).toMatchSnapshot();
  });

  test('keyboard correctly', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Option1', 'Option2', 'Option3'],
      },
    });

    const input = wrapper.find('input');
    await input.trigger('click');
    const dropdown = wrapper.findComponent({ name: 'TuSelectDropdown' });
    await input.trigger('keydown', { key: 'ArrowUp' });

    expect(dropdown.find('.is-active').text()).toBe(
      'Option3'
    );

    await input.trigger('keydown', { key: 'ArrowDown' });

    expect(dropdown.find('.is-active').text()).toBe('Option1');

    await input.trigger('keydown', { key: 'Enter' });

    expect(wrapper.emitted('change')?.[0]).toEqual(['Option1']);
  });

  test('should support mouse', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Option1', 'Option2', 'Option3'],
      },
    });

    const input = wrapper.find('input');
    await input.trigger('click');
    const dropdown = wrapper.findComponent({ name: 'TuSelectDropdown' });
    const options = dropdown.findAll('.tu-select__option');
    await options[0].trigger('mouseenter');
    await options[0].trigger('mouseleave');
    await options[1].trigger('mouseenter');
    await options[1].trigger('click');

    expect(wrapper.emitted('change')?.[0]).toEqual(['Option2']);
  });

  test('should show search option', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Option1', 'Option2', 'Option3'],
        allowSearch: true,
      },
    });

    const input = wrapper.find('input');
    await input.trigger('click');
    await input.setValue('1');
    const dropdown = wrapper.findComponent({ name: 'TuSelectDropdown' });
    const option = dropdown.find('.tu-select__option');
    expect(option.text()).toBe('Option1');
  });

  test('should enable create option', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Option1', 'Option2', 'Option3'],
        allowCreate: true,
      },
    });

    const input = wrapper.find('input');
    await input.trigger('click');
    await input.setValue('Option4');
    await input.trigger('keydown', { key: 'Enter' });

    expect(wrapper.emitted('change')?.[0]).toEqual(['Option4']);
  });
})
