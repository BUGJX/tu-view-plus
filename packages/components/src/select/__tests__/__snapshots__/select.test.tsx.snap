// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select.vue > show dropdown correctly 1`] = `"<body><div class=\\"tu-popup tu-popup--bl\\" style=\\"z-index: 1001; pointer-events: none; left: 0px; top: 10px; width: 0px;\\" trigger-placement=\\"bl\\"><div class=\\"tu-popup__wrapper tu-slide-dynamic-origin-enter-from tu-slide-dynamic-origin-enter-active\\" style=\\"transform-origin: 0 0;\\"><div class=\\"tu-popup__content\\"><div class=\\"tu-select__dropdown\\"><!--v-if--><!--v-if--><!--v-if--><div class=\\"tu-scrollbar\\"><div class=\\"tu-select__dropdown-wrap tu-scrollbar__wrap tu-scrollbar__wrap--hidden-default\\"><ul class=\\"tu-scrollbar__view tu-select__dropdown-list\\" style=\\"\\"><li class=\\"tu-select__option is-active\\"><!--v-if--><!--v-if--><span class=\\"tu-select__option-content\\">Option1</span><!--v-if--></li><li class=\\"tu-select__option\\"><!--v-if--><!--v-if--><span class=\\"tu-select__option-content\\">Option2</span><!--v-if--></li><li class=\\"tu-select__option\\"><!--v-if--><!--v-if--><span class=\\"tu-select__option-content\\">Option3</span><!--v-if--></li><li class=\\"tu-select__option\\"><!--v-if--><!--v-if--><span class=\\"tu-select__option-content\\">Option4</span><!--v-if--></li></ul><div class=\\"tu-scrollbar__bar is-horizontal\\" style=\\"display: none;\\"><div class=\\"tu-scrollbar__thumb\\" style=\\"transform: translateX(0%);\\"></div></div><div class=\\"tu-scrollbar__bar is-vertical\\" style=\\"display: none;\\"><div class=\\"tu-scrollbar__thumb\\" style=\\"transform: translateY(0%);\\"></div></div></div></div><!--v-if--></div></div><!----></div></div></body>"`;
